Adds Mutator UI and an IfElseMutator

RoboErikG opened the pull request:
This creates a field on the block specifically for the mutator config
element. This mutator field does not exist in the block's list of
inputs/fields but does create a view which is inserted into the first
input of the block's view hierarchy.

It also adds a simple mutator UI for the If Else block.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-android/552)
<!-- Reviewable:end -->

AnmAtAnm commented on blocklylib-vertical/src/main/java/com/google/blockly/android/ui/vertical/IfElseMutatorUI.java:
```diff
@@ -0,0 +1,36 @@
+import java.io.IOException;
+
+
+public class IfElseMutatorUI extends IfElseMutator {
```
Is this just a temporary test Mutator to test the UI?

RoboErikG replied:
Yes. Will replace with a real mutator once I've got the other APIs worked out.

AnmAtAnm commented on blocklylib-core/src/main/java/com/google/blockly/model/Block.java:
```diff
@@ -366,6 +368,13 @@ public void setComment(String comment) {
     /**
+     * @return The input for the mutator button if one exists.
+     */
+    public Field getMutatorField() {
```
This is only called by the BlockViewFactory, making me thing it shouldn't be part of the model.

RoboErikG commented on blocklylib-core/src/main/java/com/google/blockly/model/BlockFactory.java:
```diff
@@ -72,8 +74,12 @@
     /** Default constructor. */
-    public BlockFactory() {}
+    public BlockFactory(Context context) {
+        mMutatorAltText = context.getResources().getString(R.string.mutator_icon_alt_text);
```
All the block definitions need to be reloaded when the language changes as well. Filed #565 to track.

RoboErikG commented on blocklylib-core/src/main/java/com/google/blockly/model/FieldMutator.java:
```diff
@@ -0,0 +1,108 @@
+ * Adds the mutator edit button to an Input. This should only be added to blocks that have a mutator
+ * that {@link Mutator#hasUI() has UI}.
+ */
+public final class FieldMutator extends Field {
```
Removed the class.

RoboErikG commented on blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java:
```diff
@@ -82,6 +85,13 @@
     private static final String TAG = "AbstractBlocklyActivity";
 
+    public static final Map<String, BlockExtension> DEFAULT_EXTENSIONS = new ArrayMap<>();
+    public static final Map<String, Mutator.Factory> DEFAULT_MUTATORS = new ArrayMap<>();
```
sgtm.

RoboErikG commented:
Ready for review.

---

Review status: 0 of 44 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/552#-:-Kj5A8h_FeWKqtuBVWLF:b8pt725)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Reviewed 7 of 24 files at r1, 3 of 10 files at r2, 1 of 6 files at r3, 35 of 37 files at r4.
Review status: all files reviewed at latest revision, 33 unresolved discussions.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/TurtleActivity.java, line 36 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjEnK8k_jmpz87B_pMR:-KjEnK8k_jmpz87B_pMS:bdh1x77) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklydemo/src/main/java/com/google/blockly/android/demo/TurtleActivity.java#L36)):*
> ```Java
> import com.google.blockly.model.DefaultBlocks;
> import com.google.blockly.model.mutator.IfElseMutator;
> import com.google.blockly.model.Mutator;
> ```

Remove unused imports.

---

*[blocklylib-core/src/main/assets/default/logic_blocks.json, line 19 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFL02D28Y65-YMqZdo:-KjFL02D28Y65-YMqZdp:b-6qcu4z) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/assets/default/logic_blocks.json#L19)):*
> ```JSON
>     "nextStatement": null,
>     "colour": 210,
>     "mutator": "controls_if_mutator",
> ```

We might want to hold back on this until the reshape() is working.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 91 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjEnda0_PZ03uyFlf8m:-KjEnda0_PZ03uyFlf8n:ba3dfmt) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L91)):*
> ```Java
>         // TODO Set up default extensions and mutators
>     }
> ```

Remove.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 436 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjEoHzrEoYDq7Mp2kA8:-KjEoHzszdVM4cS9QXIv:bultfby) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L436)):*
> ```Java
>      */
>     @NonNull
>     protected Map<String, Mutator.Factory> getBlockMutators() {
> ```

Removed.  Replaced by above method.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/BlocklyActivityHelper.java, line 633 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjF9p3Mku-17VawBn2a:-KjF9p3Mku-17VawBn2b:b-fs2sdv) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/BlocklyActivityHelper.java#L633)):*
> ```Java
>             return;
>         }
>         MutatorFragment fragment = mutator.getMutatorFragment();
> ```

Can this come from the BlockViewFactory, using the mutator id, instead of from the Mutator?

I'm trying to make sure it can be extended safely, allowing app developers to match their own UI without replacing block definitions or mutators.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java, line 26 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFDjyYBy6IEEITFOtN:-KjFDjyYBy6IEEITFOtO:bp7vv6y) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java#L26)):*
> ```Java
> import android.support.annotation.Nullable;
> import android.support.annotation.VisibleForTesting;
> import android.support.v4.app.DialogFragment;
> ```

Remove. No other new changes reference this.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java, line 169 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFAXJ0g8PlS_xMjGes:-KjFAXJ0g8PlS_xMjGet:bdq5fc5) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java#L169)):*
> ```Java
>         final int inputCount = inputs.size();
>         Mutator mutator = block.getMutator();
>         boolean showMutatorUi = mutator != null && mutator.hasUI();
> ```

Replace hasUI() with a lookup of registered MutatorFragment for the mutator id. Trying to make sure the UI can be extended safely.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java, line 359 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFEQUdZmFZNPZAfRfo:-KjFEQUe3H_MFhJe0Y3e:b-n5dxrc) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java#L359)):*
> ```Java
>      * @return A {@link FieldView} for the mutator open icon.
>      */
>     protected FieldView buildMutatorFieldView(final Mutator mutator) {
> ```

TODO: Refactor so buildMutatorFieldView() can be put in VerticalBlockViewFactory.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java, line 380 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFESeeJjldlOndMfGA:-KjFESeeJjldlOndMfGB:ban7ukz) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java#L380)):*
> ```Java
>      * @return The new {@link FieldView} for the icon.
>      */
>     protected FieldView buildIconFieldView(View.OnClickListener listener, int resId) {
> ```

TODO: Refactor so buildIconFieldView() can be put in VerticalBlockViewFactory.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java, line 427 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjF8B80Ff8wuGFoZxI_:-KjF8B80Ff8wuGFoZxIa:bsw87el) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java#L427)):*
> ```Java
>      */
>     public interface MutatorToggleListener {
>         void onMutatorToggled(Mutator mutator);
> ```

I feel like this should be a block reference, since you can got from Block to Mutator, but not vice versa.

I'm not fully convinced it belongs here in the abstract BlockViewFactory instead of just the vertical-stacking view factory, but I don't think the changes required to move this are currently worth while.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java, line 1 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFKrwL0RBiRPgYJXau:-KjFKrwL0RBiRPgYJXav:b-r7l132) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java#L1)):*
> ```Java
> package com.google.blockly.android.ui;
> ```

Copyright

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java, line 7 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFKtIlTj_-SdZ-15-7:-KjFKtIlTj_-SdZ-15-8:bwywyy9) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java#L7)):*
> ```Java
> 
> 
> public abstract class MutatorFragment extends DialogFragment {
> ```

JavaDoc

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java, line 30 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFJDUXbtrOL56kBTrq:-KjFJDUXbtrOL56kBTrr:bg9uchc) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java#L30)):*
> ```Java
>      * Listener interface for performing cleanup when a MutatorFragment is dismissed.
>      */
>     public interface DismissListener {
> ```

Nit: I prefer public interfaces and nested classes at the top, since these are expected to be understood and overloaded by developers. Google Java Style doesn't seem to enforce a preference.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/WorkspaceHelper.java, line 19 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFJUgPOm6R0vIvObXu:-KjFJUgPOm6R0vIvObXv:b-bffkjb) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/WorkspaceHelper.java#L19)):*
> ```Java
> 
> import android.app.Activity;
> import android.app.Dialog;
> ```

Cleanup unused imports.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldIconView.java, line 2 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFKo_jGBVrVILk3eWg:-KjFKo_jGBVrVILk3eWh:bb77u1u) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldIconView.java#L2)):*
> ```Java
> /*
>  * Copyright 2016 Google Inc. All Rights Reserved.
> ```

Year

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java, line 1 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFKll2mGJYjvK5zbCb:-KjFKll2mGJYjvK5zbCc:b-r7l132) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java#L1)):*
> ```Java
> package com.google.blockly.android.ui.mutator;
> ```

Copyright

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java, line 23 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFKjtHHZ4h1cuPsw5k:-KjFKjtHHZ4h1cuPsw5l:b-eeqggg) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java#L23)):*
> ```Java
> import org.xmlpull.v1.XmlPullParser;
> 
> public class IfElseMutatorFragment extends MutatorFragment {
> ```

JavaDoc comment

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java, line 94 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFMKsTWDgizTK_It_a:-KjFMKsTWDgizTK_It_b:bvwpric) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java#L94)):*
> ```Java
>         });
> 
>         AlertDialog dialog = new AlertDialog.Builder(getActivity())
> ```

Refactor so the  AlertDialog construction is shared by all MutatorFragments. Relatedly, it will allow us to add support for positioned dialogs on larger screens, or other customizations.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Block.java, line 1129 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFCCJaoqABFfKSurl2:-KjFCCJaoqABFfKSurl3:b-dbrp4p) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/Block.java#L1129)):*
> ```Java
> 
>     /**
> <<<<<<< HEAD
> ```

Merge cruft

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Block.java, line 1134 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFLPHfJb5Dho7_LTVA:-KjFLPHfJb5Dho7_LTVB:b-lr307) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/Block.java#L1134)):*
> ```Java
>      * (inputs, next, previous, and output).
>      */
>     private void rebuildConnectionList() {
> ```

I don't see a call to this private method.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/BlockFactory.java, line 84 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFBnxbj3WgAQqi26rZ:-KjFBnxbj3WgAQqi26r_:b-904xf8) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/BlockFactory.java#L84)):*
> ```Java
>     /** Default constructor. */
>     public BlockFactory(Context context) {
>         mMutatorAltText = context.getResources().getString(R.string.mutator_icon_alt_text);
> ```

mutator_icon_alt_text should move to BlockViewFactory.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/BlockFactory.java, line 131 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFC08XWmXJhIK8fEX3:-KjFC08XWmXJhIK8fEX4:bv9oraf) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/BlockFactory.java#L131)):*
> ```Java
>      */
>     @VisibleForTesting BlockFactory() {
>         mMutatorAltText = "Test mutator alt text";
> ```

remove this constructor once mMutatorAltText is moved to BlockViewFactory.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/DefaultBlocks.java, line 85 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjEqNfPpYhoLO1wybuu:-KjEqNfPpYhoLO1wybuv:b-so36ya) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/DefaultBlocks.java#L85)):*
> ```Java
>      */
>     public static Map<String, Mutator.Factory> getMutators(Context context,
>             BlocklyController controller) {
> ```

Move the controller param to Mutator.Factory.newMutator(..), eliminating the need for this reference in the static map.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Field.java, line 49 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFDWASjwEnhx6R2OGy:-KjFDWATl468QnPDBq9O:brkth3t) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/Field.java#L49)):*
> ```Java
>     public static final int TYPE_IMAGE = 8;
>     public static final int TYPE_NUMBER = 9;
>     public static final int TYPE_ICON = 10;
> ```

Remove fake field type. See comment in VerticalBlockViewFactory.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java, line 43 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjF9N3EP2MRKF-L-xGr:-KjF9N3EP2MRKF-L-xGs:b-hkzosz) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java#L43)):*
> ```Java
> public abstract class Mutator {
> 
>     public static final Map<String, Factory> STANDARD_MUTATORS = Collections.EMPTY_MAP;  // TODO
> ```

Remove. Use DefaultBlocks.getMutators()

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java, line 95 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjF9CiLg5r5zlH5CyA9:-KjF9CiMUUgflGx6VJt1:b-ynys4g) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java#L95)):*
> ```Java
>      * @return true if this Mutator has an editor UI.
>      */
>     public boolean hasUI() {
> ```

Per prior comments, I don't feel the model should know if the Mutator has a UI.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java, line 61 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFEvCBvzVDIp5VtaqT:-KjFEvCBvzVDIp5VtaqU:bzn031) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java#L61)):*
> ```Java
>         mContext = context;
>         mController = controller;
>         mIfLabel = mContext.getString(R.string.mutator_if_else_if_label);
> ```

Add TODO: Replace with Blockly string table / TranslationsManager call.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java, line 110 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFHzwFgP8FLEZQ2HmY:-KjFHzwFgP8FLEZQ2HmZ:b-pi11if) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java#L110)):*
> ```Java
>     @Override
>     public MutatorFragment getMutatorFragment() {
>         IfElseMutatorFragment dialog = new IfElseMutatorFragment();
> ```

Register mutator fragment with BlockViewFaactory, and query via mutator id (so developers can easily replace the UI).

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java, line 214 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFIgI830DYj5OtEhPJ:-KjFIgI830DYj5OtEhPK:b-frfl94) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java#L214)):*
> ```Java
>         }
> 
>         mBlock.reshape(newInputs, mBlock.getOutputConnection(), mBlock.getPreviousConnection(),
> ```

There is a shorthand that doesn't take input references. Just `mBlock.reshape(newInputs);`.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java, line 226 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjEpJsqOGK2p9g6qnyo:-KjEpJsrFbh_87UcOn9Y:b-ubhrtn) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java#L226)):*
> ```Java
> 
>         public Factory(Context context, BlocklyController controller) {
>             mContext = context;
> ```

This causes DefaultBlocks.getMutators(), a static method, to require a reference to the controller and context.
Move the controller param to newMutator(), where a factory or mutator constructor can use controller.getContext().

---

*[blocklylib-vertical/src/main/java/com/google/blockly/android/ui/vertical/VerticalBlockViewFactory.java, line 173 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFD-nClY-QFxfhdubs:-KjFD-nClY-QFxfhdubt:be5fovn) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-vertical/src/main/java/com/google/blockly/android/ui/vertical/VerticalBlockViewFactory.java#L173)):*
> ```Java
>         BasicFieldIconView iconView = null;
> 
>         int layoutResId = getLayoutForField(Field.TYPE_ICON);
> ```

I think this is the only real use of Field.TYPE_ICON, and could be replaced with a separate method that returns a default_field_icon. Doing so would eliminate the need to add the fake field type.

---

*[blocklytest/src/androidTest/java/com/google/blockly/model/mutator/IfElseMutatorTest.java, line 67 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFNXgj_gC-My2YmGfm:-KjFNXgj_gC-My2YmGfn:b-48sxoo) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklytest/src/androidTest/java/com/google/blockly/model/mutator/IfElseMutatorTest.java#L67)):*
> ```Java
>                 assertThat(mBlock.getInputByName("ELSE")).isNull();
> 
>                 mutator.update(2, true);
> ```

Nit: Prefer to break up long tests, so we get more precision and coverage even when some test breaks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/552)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Review status: 19 of 47 files reviewed at latest revision, 33 unresolved discussions.

---

*[blocklydemo/src/main/java/com/google/blockly/android/demo/TurtleActivity.java, line 36 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjEnK8k_jmpz87B_pMR:-KjJVBfeJ_hR3T8AGq-n:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklydemo/src/main/java/com/google/blockly/android/demo/TurtleActivity.java#L36)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Remove unused imports.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/assets/default/logic_blocks.json, line 19 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFL02D28Y65-YMqZdo:-KjJVN0ms3wEILC00NPZ:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/assets/default/logic_blocks.json#L19)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

We might want to hold back on this until the reshape() is working.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 91 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjEnda0_PZ03uyFlf8m:-KjJVSQqb-_FpdVHSTc5:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L91)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Remove.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 436 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjEoHzrEoYDq7Mp2kA8:-KjJVaqj6OlY6A3dD_-6:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L436)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Removed.  Replaced by above method.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/BlocklyActivityHelper.java, line 633 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjF9p3Mku-17VawBn2a:-KjK9Z5V2LAHyj-ki_gq:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/BlocklyActivityHelper.java#L633)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Can this come from the BlockViewFactory, using the mutator id, instead of from the Mutator?

I'm trying to make sure it can be extended safely, allowing app developers to match their own UI without replacing block definitions or mutators.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java, line 169 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFAXJ0g8PlS_xMjGes:-KjK9epZ8Pzo-7wnF1_z:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java#L169)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Replace hasUI() with a lookup of registered MutatorFragment for the mutator id. Trying to make sure the UI can be extended safely.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java, line 359 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFEQUdZmFZNPZAfRfo:-KjOpaCZW6RAr1Z70WRz:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java#L359)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

TODO: Refactor so buildMutatorFieldView() can be put in VerticalBlockViewFactory.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java, line 380 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFESeeJjldlOndMfGA:-KjOpbIRKlh7ODCWXQ2C:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java#L380)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

TODO: Refactor so buildIconFieldView() can be put in VerticalBlockViewFactory.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java, line 427 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjF8B80Ff8wuGFoZxI_:-KjOqV2rRJxy7ovtFI8p:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/BlockViewFactory.java#L427)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I feel like this should be a block reference, since you can got from Block to Mutator, but not vice versa.

I'm not fully convinced it belongs here in the abstract BlockViewFactory instead of just the vertical-stacking view factory, but I don't think the changes required to move this are currently worth while.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java, line 1 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFKrwL0RBiRPgYJXau:-KjOqbOVftmr2XKvzuNG:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java#L1)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Copyright
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java, line 7 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFKtIlTj_-SdZ-15-7:-KjPF0nmNModj4WZzmvY:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java#L7)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

JavaDoc
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java, line 30 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFJDUXbtrOL56kBTrq:-KjPF5dbWdn2Q_IG-Eq8:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/MutatorFragment.java#L30)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Nit: I prefer public interfaces and nested classes at the top, since these are expected to be understood and overloaded by developers. Google Java Style doesn't seem to enforce a preference.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java, line 1 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFKll2mGJYjvK5zbCb:-KjPG2HXSKI3KTzmfcfG:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java#L1)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Copyright
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java, line 23 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFKjtHHZ4h1cuPsw5k:-KjPGtsQYX_f4QF0_wkd:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java#L23)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

JavaDoc comment
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java, line 94 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFMKsTWDgizTK_It_a:-KjPH1AprbrtavTPlzC8:bzztvl) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java#L94)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Refactor so the  AlertDialog construction is shared by all MutatorFragments. Relatedly, it will allow us to add support for positioned dialogs on larger screens, or other customizations.
</blockquote></details>

That won't simplify the API at all since I still need to pass the title and content view through. Also, AlertDialogs are a bit restrictive and I'd rather not make them the default.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Block.java, line 1129 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFCCJaoqABFfKSurl2:-KjPHTOCwKMN0G7H2dph:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/Block.java#L1129)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Merge cruft
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Block.java, line 1134 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFLPHfJb5Dho7_LTVA:-KjPHRFQGBNF05q1CLSE:b-niophj) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/Block.java#L1134)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I don't see a call to this private method.
</blockquote></details>

I think this was leftover from a merge. Removed.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/BlockFactory.java, line 467 at r1](https://reviewable.io:443/reviews/google/blockly-android/552#-KgoyTBkGfJOhQIFzjb6-r1-467:-KjPIfuit6ugAqF1Wkdp:bnvl6hs) ([raw file](https://github.com/google/blockly-android/blob/3a5cc471a317fbd5f4a7e546c57cf34692d9c553/blocklylib-core/src/main/java/com/google/blockly/model/BlockFactory.java#L467)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Does the mutator field have to exist at the model level? I thought of it's creation (and placement) to be specific to the BlockView implementation.  As is, you are specifying sizing and icon assets here in the BlockFactory.
</blockquote></details>

mutator field has been removed. Let me know if I've missed some UI cleanup.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/BlockFactory.java, line 84 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFBnxbj3WgAQqi26rZ:-KjPIAFn_qhUdiUBdtaK:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/BlockFactory.java#L84)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

mutator_icon_alt_text should move to BlockViewFactory.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/BlockFactory.java, line 131 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFC08XWmXJhIK8fEX3:-KjPIWIttIdMttdyl6Hl:bbcl1ms) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/BlockFactory.java#L131)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

remove this constructor once mMutatorAltText is moved to BlockViewFactory.
</blockquote></details>

actually, made the one without a context the default again.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/DefaultBlocks.java, line 85 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjEqNfPpYhoLO1wybuu:-KjPJcadz578O0IGLgUT:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/DefaultBlocks.java#L85)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Move the controller param to Mutator.Factory.newMutator(..), eliminating the need for this reference in the static map.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Field.java, line 49 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFDWASjwEnhx6R2OGy:-KjPKB18oLjASIzJQsCh:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/Field.java#L49)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Remove fake field type. See comment in VerticalBlockViewFactory.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java, line 43 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjF9N3EP2MRKF-L-xGr:-KjPKFF4-F1t7BPk2wfp:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java#L43)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Remove. Use DefaultBlocks.getMutators()
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java, line 95 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjF9CiLg5r5zlH5CyA9:-KjPKINKG9-skPLRfv1b:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java#L95)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Per prior comments, I don't feel the model should know if the Mutator has a UI.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java, line 61 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFEvCBvzVDIp5VtaqT:-KjPKQg3VwKCortiqwn4:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java#L61)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Add TODO: Replace with Blockly string table / TranslationsManager call.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java, line 110 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFHzwFgP8FLEZQ2HmY:-KjPKUdpg5kcYNzj5JeK:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java#L110)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Register mutator fragment with BlockViewFaactory, and query via mutator id (so developers can easily replace the UI).
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java, line 214 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFIgI830DYj5OtEhPJ:-KjPLK5bG7kyT1u3FO7z:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java#L214)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

There is a shorthand that doesn't take input references. Just `mBlock.reshape(newInputs);`.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java, line 226 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjEpJsqOGK2p9g6qnyo:-KjPKdWYZ0Jd2EsHuY2S:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/model/mutator/IfElseMutator.java#L226)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

This causes DefaultBlocks.getMutators(), a static method, to require a reference to the controller and context.
Move the controller param to newMutator(), where a factory or mutator constructor can use controller.getContext().
</blockquote></details>

Done.

---

*[blocklylib-vertical/src/main/java/com/google/blockly/android/ui/vertical/VerticalBlockViewFactory.java, line 173 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFD-nClY-QFxfhdubs:-KjPLMiZCapfcMO2NG4m:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-vertical/src/main/java/com/google/blockly/android/ui/vertical/VerticalBlockViewFactory.java#L173)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

I think this is the only real use of Field.TYPE_ICON, and could be replaced with a separate method that returns a default_field_icon. Doing so would eliminate the need to add the fake field type.
</blockquote></details>

Done.

---

*[blocklytest/src/androidTest/java/com/google/blockly/model/mutator/IfElseMutatorTest.java, line 67 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFNXgj_gC-My2YmGfm:-KjPMGb-OHWgFxLn8nM6:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklytest/src/androidTest/java/com/google/blockly/model/mutator/IfElseMutatorTest.java#L67)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Nit: Prefer to break up long tests, so we get more precision and coverage even when some test breaks.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java, line 26 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFDjyYBy6IEEITFOtN:-KjK9dBPc18akB_uakWk:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/control/BlocklyController.java#L26)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Remove. No other new changes reference this.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/WorkspaceHelper.java, line 19 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFJUgPOm6R0vIvObXu:-KjPFEMunFpBfkkbDXsC:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/WorkspaceHelper.java#L19)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Cleanup unused imports.
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldIconView.java, line 2 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFKo_jGBVrVILk3eWg:-KjPFz0ioQ1nae0QJNOn:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/fieldview/BasicFieldIconView.java#L2)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Year
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/552)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Reviewed 30 of 32 files at r5.
Review status: 41 of 47 files reviewed at latest revision, 5 unresolved discussions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 88 at r5](https://reviewable.io:443/reviews/google/blockly-android/552#-KjPp62mI-InjmWtXsbl:-KjPp62mI-InjmWtXsbm:b-6iiuj3) ([raw file](https://github.com/google/blockly-android/blob/d55e44928d3338259377ad7992b36d9099a3dbe7/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L88)):*
> ```Java
>     private static final String TAG = "AbstractBlocklyActivity";
> 
>     protected BlocklyActivityHelper mActivityHelper;
> ```

Do we really want to do this? The naming mBlockly has been in the public sphere for 3 months.  At the very least, it should be a separate commit.

I originally considered mHelper, but that is already often used to a refer to a WorkspaceHelper. Even so, I prefere mBlocklyHelper, because this is the thing other developers will drop into their non-Blockly code, so  "Blockly" should be in the name.

(Okay.  This may be stretching the rationale for the field name, compared to the class name, but makes it easier to cut and paste AbstractBlocklyActivity code into an existing activity if we assume the field name is the same.)

---

*[blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java, line 94 at r4](https://reviewable.io:443/reviews/google/blockly-android/552#-KjFMKsTWDgizTK_It_a:-Kjcumr2axiJ-Z1SG-z5:bkrgp8i) ([raw file](https://github.com/google/blockly-android/blob/c9d5967dea9470ef34605f890b0d883f3487a0d0/blocklylib-core/src/main/java/com/google/blockly/android/ui/mutator/IfElseMutatorFragment.java#L94)):*
<details><summary><i>Previously, RoboErikG wrote…</i></summary><blockquote>

That won't simplify the API at all since I still need to pass the title and content view through. Also, AlertDialogs are a bit restrictive and I'd rather not make them the default.
</blockquote></details>

Fine for now, but I'm concerned we are leaving the sizing positioning left to each MutatorFragment.

My preference would be to have a Fragment that is inserted into a frame managed by the.... BlockViewFactory? (Too many things going into that, but some similar UI singleton on the UI side.) Then, there is a single place to control placement (especially on larger screen real estate devices), and a single place to tweak the mutator dialog chrome.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java, line 67 at r5](https://reviewable.io:443/reviews/google/blockly-android/552#-Kjczpn0S2hClpNX7Fly:-Kjczpn0S2hClpNX7Flz:b60icd3) ([raw file](https://github.com/google/blockly-android/blob/d55e44928d3338259377ad7992b36d9099a3dbe7/blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java#L67)):*
> ```Java
>     }
> 
>     public abstract String getMutatorId();
> ```

Per offline, make this a final implementation, add abstract version to Factory, and add protected constructor Mutator(Factory).

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 20 at r5](https://reviewable.io:443/reviews/google/blockly-android/552#-Kjd7Q8mBdChNqFvp0F-:-Kjd7Q8mBdChNqFvp0F0:buzc58k) ([raw file](https://github.com/google/blockly-android/blob/d55e44928d3338259377ad7992b36d9099a3dbe7/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L20)):*
> ```Java
> import android.support.annotation.NonNull;
> import android.support.annotation.Nullable;
> import android.support.annotation.VisibleForTesting;
> ```

unused?

---

*[blocklytest/src/androidTest/java/com/google/blockly/model/MutatorTest.java, line 130 at r5](https://reviewable.io:443/reviews/google/blockly-android/552#-Kjd7xcnetfGz6gXsLJe:-Kjd7xco0RAE0n7yIfjf:bab6d94) ([raw file](https://github.com/google/blockly-android/blob/d55e44928d3338259377ad7992b36d9099a3dbe7/blocklytest/src/androidTest/java/com/google/blockly/model/MutatorTest.java#L130)):*
> ```Java
> 
>         @Override
>         public String getMutatorId() {
> ```

Make sure this gets removed (replaced by default implementation).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/552)*
<!-- Sent from Reviewable.io -->

RoboErikG commented:
Review status: 37 of 45 files reviewed at latest revision, 5 unresolved discussions.

---

*[blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java, line 88 at r5](https://reviewable.io:443/reviews/google/blockly-android/552#-KjPp62mI-InjmWtXsbl:-KjdG1AUlc30bA0Nq8ru:bxrfpmy) ([raw file](https://github.com/google/blockly-android/blob/d55e44928d3338259377ad7992b36d9099a3dbe7/blocklylib-core/src/main/java/com/google/blockly/android/AbstractBlocklyActivity.java#L88)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Do we really want to do this? The naming mBlockly has been in the public sphere for 3 months.  At the very least, it should be a separate commit.

I originally considered mHelper, but that is already often used to a refer to a WorkspaceHelper. Even so, I prefere mBlocklyHelper, because this is the thing other developers will drop into their non-Blockly code, so  "Blockly" should be in the name.

(Okay.  This may be stretching the rationale for the field name, compared to the class name, but makes it easier to cut and paste AbstractBlocklyActivity code into an existing activity if we assume the field name is the same.)
</blockquote></details>

undid the rename. Will follow up in a separate CL to change it to mBlocklyActivityHelper to make it clearer what it is.

---

*[blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java, line 67 at r5](https://reviewable.io:443/reviews/google/blockly-android/552#-Kjczpn0S2hClpNX7Fly:-KjdGj4-hLvzrpCNudqV:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/d55e44928d3338259377ad7992b36d9099a3dbe7/blocklylib-core/src/main/java/com/google/blockly/model/Mutator.java#L67)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Per offline, make this a final implementation, add abstract version to Factory, and add protected constructor Mutator(Factory).
</blockquote></details>

Done.

---

*[blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java, line 20 at r5](https://reviewable.io:443/reviews/google/blockly-android/552#-Kjd7Q8mBdChNqFvp0F-:-KjdGobqRx2hUoTk76O6:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/d55e44928d3338259377ad7992b36d9099a3dbe7/blocklylib-core/src/main/java/com/google/blockly/utils/BlocklyXmlHelper.java#L20)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

unused?
</blockquote></details>

Done.

---

*[blocklytest/src/androidTest/java/com/google/blockly/model/MutatorTest.java, line 130 at r5](https://reviewable.io:443/reviews/google/blockly-android/552#-Kjd7xcnetfGz6gXsLJe:-KjdGrVDibSHsF0Lb0cd:b-896fix) ([raw file](https://github.com/google/blockly-android/blob/d55e44928d3338259377ad7992b36d9099a3dbe7/blocklytest/src/androidTest/java/com/google/blockly/model/MutatorTest.java#L130)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Make sure this gets removed (replaced by default implementation).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/552)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 10 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-android/552#-:-KjdI2qbE9fezCQ9KsxN:bnfp4nl)*
<!-- Sent from Reviewable.io -->
